stm32: eth: build issue after PHY address resolution fix (84889d4)

### Discussed in https://github.com/zephyrproject-rtos/zephyr/discussions/77662

<div type='discussions-op-text'>

<sup>Originally posted by **scaprile** August 27, 2024</sup>
The following boards stopped building after https://github.com/zephyrproject-rtos/zephyr/commit/84889d478362876e3a2bf4669e205def14c6794b:

nucleo_h563zi
stm32h573i_dk
nucleo_h743zi
stm32h735g_disco
stm32h745i_disco/stm32h745xx/m7
stm32h747i_disco/stm32h747xx/m7
nucleo_h753zi
nucleo_h755zi_q/stm32h755xx/m7

The error is this one:
```
zephyr/drivers/ethernet/eth_stm32_hal.c:62:92: error: pasting ")" and "_ORD" does not give a valid preprocessing token
   62 |             DEVICE_DT_GET(DT_CHILD(DT_INST_CHILD(n, mdio), __CONCAT(ethernet_phy_, PHY_ADDR)))
```
looks like `CONFIG_ETH_STM32_HAL_PHY_ADDRESS` does not get a value (hence also `PHY_ADDR`)

**`nucleo_h563zi`, for example, builds again when rolling back that commit.**

My prj.conf is bare minimum
```
CONFIG_NETWORKING=y
CONFIG_NET_IPV4=y
CONFIG_NET_IPV6=y
CONFIG_NET_TCP=y
CONFIG_NET_UDP=y
CONFIG_NET_DHCPV4=y
CONFIG_NET_SOCKETS=y
CONFIG_NET_SOCKETS_POLL_MAX=32
CONFIG_POSIX_API=y
CONFIG_POSIX_MAX_FDS=32
CONFIG_NET_MAX_CONN=10
CONFIG_NET_MAX_CONTEXTS=10
CONFIG_NET_CONFIG_SETTINGS=y
CONFIG_NET_CONNECTION_MANAGER=y
CONFIG_NET_LOG=y

CONFIG_LOG=y
CONFIG_ISR_STACK_SIZE=2048
CONFIG_MAIN_STACK_SIZE=8192
CONFIG_IDLE_STACK_SIZE=1024

CONFIG_MINIMAL_LIBC=y
CONFIG_MINIMAL_LIBC_RAND=y
CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=32768
```
Is this an error on my side ? Something essential is missing that breaks Kconfig reading ?

These boards, nevertheless, build OK with that very same prj.conf:

nucleo_f207zg
nucleo_f429zi
nucleo_f746zg
nucleo_f767zi
</div>

@AlexFabre `stm32h745i_disco/stm32h745xx/m7` has its PHY hardwired for address 1:
![Screenshot from 2024-08-29 16-08-42](https://github.com/user-attachments/assets/8e1f5a22-57a9-4c7d-b6b5-784d8bdd92ad)
I would expect Zephyr to configure for that board without user intervention; though I see `CONFIG_ETH_STM32_HAL_PHY_ADDRESS=0` in `build/zephyr/.config`.
Technically, one could force that PHY pin low on reset by setting the GPIO to an output low, releasing reset, and then configuring as an input; I don't know if you folks do something like that, I hope you don't, and I don't currently have a board with me to test if it actually works.

Setting `CONFIG_ETH_STM32_HAL_PHY_ADDRESS=1` in `prj.conf` breaks the build, now the error is this one:
```
                 from zephyr/drivers/ethernet/eth_stm32_hal.c:13:
zephyr/include/zephyr/device.h:92:41: error: '__device_dts_ord_DT_N_S_soc_S_ethernet_40028000_S_mdio_S_ethernet_phy_1_ORD' undeclared here (not in a function); did you mean 'DT_N_S_soc_S_ethernet_40028000_S_mdio_S_ethernet_phy_0_ORD'?
   92 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
zephyr/include/zephyr/toolchain/common.h:137:26: note: in definition of macro '_DO_CONCAT'
  137 | #define _DO_CONCAT(x, y) x ## y
      |                          ^
zephyr/include/zephyr/device.h:92:33: note: in expansion of macro '_CONCAT'
   92 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                 ^~~~~~~
```
Please note that not adding that line, or even setting `CONFIG_ETH_STM32_HAL_PHY_ADDRESS=0` builds OK.
Let me know if you would like this to be in a separate issue.

--- a/drivers/ethernet/eth_stm32_hal.c
+++ b/drivers/ethernet/eth_stm32_hal.c
@@ -59,7 +59,7 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #if defined(CONFIG_MDIO)
 
 #define DEVICE_PHY_BY_NAME(n) \
-	    DEVICE_DT_GET(DT_CHILD(DT_INST_CHILD(n, mdio), __CONCAT(ethernet_phy_, PHY_ADDR)))
+	    DEVICE_DT_GET(DT_CHILD(DT_INST_CHILD(n, mdio), _CONCAT(ethernet_phy_, PHY_ADDR)))
 
 static const struct device *eth_stm32_phy_dev = DEVICE_PHY_BY_NAME(0);
 

